compiler/qsc_qasm3/src/compile.rs

```diff
@@ -0,0 +1,4283 @@
+                let mut existing_attrs = item.attrs.to_vec();
+                existing_attrs.extend(current_annotations);
+                item.attrs = existing_attrs.into_boxed_slice();
+            } else {
+                let kind = SemanticErrorKind::InvalidAnnotationTarget(stmt.span);
+                self.push_semantic_error(kind);
+            }
+        } else {
+            let kind = SemanticErrorKind::InvalidAnnotationTarget(stmt.span);
+            self.push_semantic_error(kind);
+        }
+    }
+
+    /// Match against the different types of statements and compile them
+    /// to the appropriate AST statement. There should be no logic here.
+    fn compile_stmt(&mut self, stmt: &oq3_syntax::ast::Stmt) -> Option<ast::Stmt> {
+        match stmt {
+            Stmt::AliasDeclarationStatement(alias) => self.compile_alias_decl(alias),
+            Stmt::AssignmentStmt(assignment) => self.compile_assignment_stmt(assignment),
+            Stmt::Barrier(barrier) => self.compile_barrier_stmt(barrier),
+            Stmt::BreakStmt(break_stmt) => self.compile_break_stmt(break_stmt),
+            Stmt::ClassicalDeclarationStatement(decl) => self.compile_classical_decl(decl),
+            Stmt::ContinueStmt(continue_stmt) => self.compile_continue_stmt(continue_stmt),
+            Stmt::Def(def) => self.compile_def_decl(def),
+            Stmt::EndStmt(end) => Some(compile_end_stmt(end)),
+            Stmt::ExprStmt(expr) => self.compile_expr_stmt(expr),
+            Stmt::ForStmt(for_stmt) => self.compile_for_stmt(for_stmt),
+            Stmt::Gate(gate) => self.compile_gate_decl(gate),
+            Stmt::IfStmt(if_stmt) => self.compile_if_stmt(if_stmt),
+            Stmt::IODeclarationStatement(io_decl) => self.compile_io_decl_stmt(io_decl),
+            Stmt::LetStmt(let_stmt) => self.compile_let_stmt(let_stmt),
+            Stmt::Measure(measure) => self.compile_measure_stmt(measure),
+            Stmt::QuantumDeclarationStatement(decl) => self.compile_quantum_decl(decl),
+            Stmt::Reset(reset) => self.compile_reset_call(reset),
+            Stmt::SwitchCaseStmt(switch_case) => self.compile_switch_stmt(switch_case),
+            Stmt::VersionString(version) => self.compile_version_stmt(version),
+            Stmt::WhileStmt(while_stmt) => self.compile_while_stmt(while_stmt),
+            Stmt::Include(include) => self.compile_include_stmt(include),
+            Stmt::Cal(..) | Stmt::DefCal(..) | Stmt::DefCalGrammar(..) => {
+                self.compile_calibration_stmt(stmt)
+            }
+            Stmt::AnnotationStatement(annotation) => {
+                let message = "Annotation statements should have been handled in compile_stmts";
+                let span = span_for_syntax_node(annotation.syntax());
+                let kind = SemanticErrorKind::UnexpectedParserError(message.to_string(), span);
+                self.push_semantic_error(kind);
+                None
+            }
```

Comment:
While this is friendlier than a panic, this seems like the kind of place I'd expecte to see an explicit panic. Rather than getting here because of some error in the parser, this actually happens only if the logic in `compile_stmts` changes in a way that breaks this assumption. I think it's worth considering changing this to an explicit panic.